### DISCUSSION TITLE
Hops: experimental Homebrew bottle installer and relocator

### DISCUSSION THREAD

**Discussion by justenstall** [Homebrew/discussions#5460] [Everyday usage]
Hello! I have been working on an experimental Homebrew bottle installer called Hops: https://github.com/act3-ai/hops

The goal was to use Homebrew in air gapped networks, which is enabled by a command that copies bottles and their metadata to a standalone OCI registry and an install command that uses the standalone registry. The project was written in Go, so I was able to use concurrency to improve performance, especially for downloads.

Would love to share more about the experiment if there is any interest
@MikeMcQuaid

**Comment by MikeMcQuaid** [comment]
@justenstall This sounds interesting, thanks for sharing. Could you provide more details on how this works for the `hops` runner for the end-user? What's your personal use-case here? Thanks!

**Reply by justenstall** [reply]
Our current scenario is:

- We have an air gapped network (no outside internet connection)
- We need to use various tools within the network
- We produce a `.deb` package containing all of the necessary tools
- The `.deb` package is uploaded to a package registry in the air gapped network so developers can install the tools there

We would like to use Homebrew for this because:

- There are Homebrew Formulae for almost all of the tools we are using (but there are not always existing Debian packages)
- Homebrew is cross-platform for macOS and Linux (current approach only works for Linux)
- We already use Homebrew to have our developers install internal tools from a private registry (with a custom `DownloadStrategy`): [act3-ai/tap](https://github.com/act3-ai/homebrew-tap)

Hops lets us use Homebrew bottles in the air gapped network with an OCI registry as the only requirement.

With Hops:

1. The administrator of the air gapped network runs [`hops copy`](https://github.com/act3-ai/hops/blob/main/docs/cli/copy.md) to copy all necessary bottles and their metadata to a destination OCI registry
   - The command takes a list of formulae to copy
3. The contents of that OCI registry are uploaded to the air gapped network's OCI registry
4. End users (in the air gapped network) configure their Hops CLI to use the OCI registry as the single source of truth for bottles/metadata
   - Set by `--registry` command line flag, `HOPS_REGISTRY` environment variable, or `.registry.prefix` in Hops' YAML config file
   - Works similar to the behavior of `HOMEBREW_BOTTLE_DOMAIN`, `HOMEBREW_ARTIFACT_DOMAIN`, and `HOMEBREW_API_DOMAIN` but expects all content to be available in the registry
6. When an end user runs `hops install ___`, the formula is installed as normal, but the API data and bottle file are downloaded from the same registry

That's a lot of information, let me know if there's anything I can clarify.

**Reply by MikeMcQuaid** [reply]
Thanks @justenstall! A few more questions.

> 2\. The contents of that OCI registry are uploaded to the air gapped network's OCI registry

How does this occur? Is `hops` used here or other standard Unix tools?

> 4\. When an end user runs `hops install ___`, the formula is installed as normal, but the API data and bottle file are downloaded from the same registry

Does `hops install` basically do "`brew fetch` but from `HOPS_REGISTRY`" and then "`brew install` as normal from Homebrew but relying on the cached files?

> * Works similar to the behavior of `HOMEBREW_BOTTLE_DOMAIN`, `HOMEBREW_ARTIFACT_DOMAIN`, and `HOMEBREW_API_DOMAIN` but expects all content to be available in the registry

Out of interest: what is it that made you decide to create a new tool rather than use these variables? Another way of asking, perhaps: what about these variables was too limited for your use-case?

**Reply by justenstall** [reply]
> > 2. The contents of that OCI registry are uploaded to the air gapped network's OCI registry
> 
> How does this occur? Is `hops` used here or other standard Unix tools?

Other tools are used for this.

> > 4. When an end user runs `hops install ___`, the formula is installed as normal, but the API data and bottle file are downloaded from the same registry
> 
> Does `hops install` basically do "`brew fetch` but from `HOPS_REGISTRY`" and then "`brew install` as normal from Homebrew but relying on the cached files?

Not quite. `brew fetch` and `brew install` rely on the `formulae.brew.sh` API (or the `homebrew-core` Git repository if the API is disabled). `hops install` (when using `HOPS_REGISTRY`) does not rely on the `formulae.brew.sh` API, because the API data is uploaded to the registry.

> > * Works similar to the behavior of `HOMEBREW_BOTTLE_DOMAIN`, `HOMEBREW_ARTIFACT_DOMAIN`, and `HOMEBREW_API_DOMAIN` but expects all content to be available in the registry
> 
> Out of interest: what is it that made you decide to create a new tool rather than use these variables? Another way of asking, perhaps: what about these variables was too limited for your use-case?

To use the existing variables, we would have had to mirror two data sources:

1. Either the `formulae.brew.sh` API or the `homebrew-core` Git repository
2. The contents of the `ghcr.io/homebrew/core` container registry

That could have worked, but it was theoretically possible the mirrored data sources could get out of sync (also we did not want to host a mirror of the API), so we came up with a way to bundle the API data for each formula with the bottle files in the container registry so the bottles are self describing. The rest of `hops` was only necessary because `brew` doesn't know how to consume the self describing bottles.

**Reply by MikeMcQuaid** [reply]
> Other tools are used for this.

Are you willing/able to detail these tools or is this your "secret sauce"? No worries if so, just interested in figuring out how the whole workflow goes here.

> `hops install` (when using `HOPS_REGISTRY`) does not rely on the `formulae.brew.sh` API, because the API data is uploaded to the registry.

So is Homebrew not doing e.g. relocation of files after installation and the usual preinstall/postinstall checks provided by `brew install`?

> so we came up with a way to bundle the API data for each formula with the bottle files in the container registry so the bottles are self describing.

Nice. This is something we'd consider upstreaming, if you were interested in either working on this or detailing the approach so someone else could?

**Reply by justenstall** [reply]
> Are you willing/able to detail these tools or is this your "secret sauce"? No worries if so, just interested in figuring out how the whole workflow goes here.

We use an internal tool for mirroring OCI images/artifacts. It is maintained on our private Git host but there is an open source mirror on GitLab.com: https://gitlab.com/act3-ai/asce/data/tool

> So is Homebrew not doing e.g. relocation of files after installation and the usual preinstall/postinstall checks provided by `brew install`?

No, Hops does not depend on Homebrew. I have reimplemented the fetch, pour, link, etc. functionality in `hops install`.

> Nice. This is something we'd consider upstreaming, if you were interested in either working on this or detailing the approach so someone else could?

Absolutely! Let me know how you'd like to start that process. I can open an issue, but meeting to discuss the details beforehand might be helpful.

**Reply by MikeMcQuaid** [reply]
> We use an internal tool for mirroring OCI images/artifacts. It is maintained on our private Git host but there is an open source mirror on GitLab.com: [gitlab.com/act3-ai/asce/data/tool](https://gitlab.com/act3-ai/asce/data/tool)

Ok, thanks!

> No, Hops does not depend on Homebrew. I have reimplemented the fetch, pour, link, etc. functionality in `hops install`.

Ok. Do users have Homebrew installed at all? If they do: I'd appreciate if you could write something to e.g. the `INSTALL_RECEIPT` or Cellar or keg or similar to indicate this. It's almost inevitable that there will be bugs at some point from the differences in pouring/linking behaviour and it'd be nice for Homebrew maintainers to not have to debug this for folks who might not be aware that Hops is the problem. Please also ensure Hops users report issues with these sort of things to you rather than us.

> Absolutely! Let me know how you'd like to start that process. I can open an issue, but meeting to discuss the details beforehand might be helpful.

Better to just do it in an issue. Meetings are unnecessarily hard for me to schedule nowadays.

Happy to discuss here, though. What are the parts you are most interested in upstreaming?

**Reply by justenstall** [reply]
> Do users have Homebrew installed at all?

No, and I will make sure error messages direct to the Hops repo instead of Homebrew.

**Reply by justenstall** [reply]
> What are the parts you are most interested in upstreaming?

I'd like to upstream the bottle registry changes including the self-contained bottle images and pushing `latest` tags so the bottle version can be determined without the API.

If there is any interest in rewriting Homebrew's CLI in Go, Hops is a good starting point for what that would look like. There are some performance, maintainability, and distribution benefits that I think would make it worth looking into. Since Homebrew's commands are already a mix of Ruby and shell scripts, it would be well suited to a progressive migration if it got to that point.

**Reply by MikeMcQuaid** [reply]
> No, and I will make sure error messages direct to the Hops repo instead of Homebrew.

Thanks. It'd still be good if Homebrew and Hops both share an installation location to write something specifically to `INSTALL_RECEIPT.json` or have a `#{keg}/.hopsmetadata` file or similar (if you don't already, if you do: please let us know!) so we can detect and handle this on our side, too.

I guess I have a question about how you handle things like e.g. `def post_install` blocks without Homebrew installed? Are they just skipped for now?

**Reply by MikeMcQuaid** [reply]
> I'd like to upstream the bottle registry changes including the self-contained bottle images and pushing `latest` tags so the bottle version can be determined without the API.

I like this idea but, if you don't use the API, how are you doing checksum validation here?

**Reply by MikeMcQuaid** [reply]
> If there is any interest in rewriting Homebrew's CLI in Go, Hops is a good starting point for what that would look like. There are some performance, maintainability, and distribution benefits that I think would make it worth looking into. Since Homebrew's commands are already a mix of Ruby and shell scripts, it would be well suited to a progressive migration if it got to that point.

Homebrew itself could never really fully be moved to Go without breaking the Formula DSL (which heavily relies on Ruby syntax/standard library) but I agree it would be an interesting idea to explore improving the performance of Homebrew by narrowing the scope to "only works for installing bottles from JSON into the supported location" using a Hops-like approach.

**Reply by justenstall** [reply]
> I like this idea but, if you don't use the API, how are you doing checksum validation here?

As long as I trust that the version tag is pointing to the correct manifest, then I know that the bottle and API JSON referenced in the tree are valid and I can verify their digests/checksums when downloading either.

Signing artifacts would guarantee a mirror of the registry couldn't change the tags maliciously (back to [this issue](https://github.com/Homebrew/brew/issues/17838) again!)

**Reply by justenstall** [reply]
> I guess I have a question about how you handle things like e.g. `def post_install` blocks without Homebrew installed? Are they just skipped for now?

> Homebrew itself could never really fully be moved to Go without breaking the Formula DSL (which heavily relies on Ruby syntax/standard library) but I agree it would be an interesting idea to explore improving the performance of Homebrew by narrowing the scope to "only works for installing bottles from JSON into the supported location" using a Hops-like approach.

Yep, using Ruby as the Formula DSL defines the limit of what could be easily moved. If the "run the Formula DSL file" Ruby code was an isolated component, I would like to know how often a consumer would need to use it. Since it's only used for installing from source and the `post_install` steps, I know that my day-to-day use of Homebrew does not require that component except when I am installing from a third-party tap.

If the Formula DSL runner was its own Ruby program, an updated CLI implementation could call the Ruby program when needed. That would have the same performance benefits and most of the maintainability benefits, but it would still rely on a Ruby interpreter at times.

**Reply by MikeMcQuaid** [reply]
There's definitely some more code that could be moved to Go for more users in those cases, agreed. Unfortunately, it would require breaking far more backwards compatibility that we're willing to do for a long while. Never say never, though!

Thanks for the fruitful conversation, @justenstall.
